lib/sse_parser.ex

```diff
@@ -0,0 +1,151 @@
+  end
+
+  defmodule State do
+    defstruct listeners: MapSet.new(), event: %Event{}, line_part: ""
+  end
+
+
+  ## Client API
+
+  @doc """
+  Starts a new SSEParser.
+  """
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, :ok, opts)
+  end
+
+  @doc """
+  Feeds some input data into the parser.
+  """
+  def feed_data(server, data) do
+    GenServer.cast(server, {:feed_data, data})
+  end
+
+  @doc """
+  Registers a listener process to receive parsed events.
+  """
+  def register_listener(server, pid) do
+    GenServer.call(server, {:register_listener, pid})
+  end
+
+  @doc """
+  Unregisters a listener process.
+  """
+  def unregister_listener(server, pid) do
+    GenServer.call(server, {:unregister_listener, pid})
+  end
+
+
+  ## Server callbacks
+
+  def init(:ok) do
+    {:ok, %State{}}
+  end
+
+  def handle_cast({:feed_data, data}, state) do
+    new_state = data_received(data, state)
+    {:noreply, new_state}
+  end
+
+  def handle_call({:register_listener, pid}, _from, state) do
+    new_listeners = MapSet.put(state.listeners, pid)
+    {:reply, :ok, %{state | listeners: new_listeners}}
+  end
+
+  def handle_call({:unregister_listener, pid}, _from, state) do
+    new_listeners = MapSet.delete(state.listeners, pid)
+    {:reply, :ok, %{state | listeners: new_listeners}}
+  end
+
+  def handle_call(:_get_state, _from, state) do
+    {:reply, {:ok, state}, state}
+  end
+
+
+  ## Internals
+
+  def emit_event(event, state) do
+    Enum.each(state.listeners, fn l -> send(l, {:sse, event}) end)
+  end
+
+
+  ## Parser
+
+  # This clause handles the end of the input.
+  defp data_received("", state) do
+    state
+  end
+  # These three clauses handle newlines.
+  defp data_received("\r\n" <> data, state), do: line_complete(data, state)
+  defp data_received("\r" <> data, state), do: line_complete(data, state)
+  defp data_received("\n" <> data, state), do: line_complete(data, state)
+  # This clause handles anything not matche above, which is all non-newlines
+  # characters.
+  defp data_received(<<char, data :: binary>>, state) do
+    %State{line_part: line} = state
+    new_state = %{state | line_part: line <> <<char>>}
+    data_received(data, new_state)
+  end
+
+  defp line_complete(data, state) do
+    new_state = line_received(state.line_part, %{state | line_part: ""})
+    data_received(data, new_state)
+  end
+
+  # Handle an empty line, which indicates the end of an event.
+  defp line_received("", state) do
+    if state.event.data != "" do
+      # Remove one trailing newline (if there is one).
+      data = String.replace_suffix(state.event.data, "\n", "")
+      emit_event(%{state.event | data: data}, state)
+    end
+    %{state | event: %Event{}}
+  end
+  # Handle a comment by ignoring it.
+  defp line_received(":" <> _, state), do: state
+  # Parse the line into the field and value for further processing.
+  defp line_received(line, state) do
+    {field, value} = case String.split(line, ":") do
```

Comment:
I _think_ you need `String.split(line, ":", parts: 2)` and a test case for that.

This case is pretty cool (or at least much nicer than Python) otherwise :)

Reply:
Good point. I'll write a test and fix that.